server/src/main/java/timely/Configuration.java

```diff
@@ -1,196 +1,511 @@
-        props.setProperty(VISIBILITY_CACHE_MAX_CAPACITY, VISIBILITY_CACHE_MAX_CAPACITY_DEFAULT + "");
-
-        props.setProperty(WS_TIMEOUT_SECONDS, WS_TIMEOUT_SECONDS_DEFAULT);
-        props.setProperty(WS_SUBSCRIPTION_LAG, WS_SUBSCRIPTION_LAG_DEFAULT);
-
-        props.setProperty(NON_SECURE_REDIRECT_PATH, NON_SECURE_REDIRECT_PATH_DEFAULT);
-        props.setProperty(STRICT_TRANSPORT_MAX_AGE, STRICT_TRANSPORT_MAX_AGE_DEFAULT);
-
-        try (BufferedReader reader = new BufferedReader(new InputStreamReader(configStream, StandardCharsets.UTF_8))) {
-            props.load(reader);
-            validate();
-        }
-    }
-
-    protected void validate() {
-        REQUIRED_PROPERTIES.forEach(p -> {
-            if (!props.containsKey(p) || StringUtils.isEmpty(props.getProperty(p))) {
-                throw new IllegalArgumentException("Required property " + p + " must be specified.");
-            }
-        });
-        if (!getBoolean(Configuration.SSL_USE_GENERATED_KEYPAIR)) {
-            if (StringUtils.isEmpty(get(Configuration.SSL_CERTIFICATE_FILE))) {
-                throw new IllegalArgumentException(Configuration.SSL_CERTIFICATE_FILE + "must be specified.");
-            }
-            if (StringUtils.isEmpty(get(Configuration.SSL_PRIVATE_KEY_FILE))) {
-                throw new IllegalArgumentException(Configuration.SSL_PRIVATE_KEY_FILE + "must be specified.");
-            }
-        }
-    }
-
-    public String get(String property) {
-        return props.getProperty(property);
-    }
-
-    public boolean getBoolean(String property) {
-        return Boolean.parseBoolean(get(property));
+    @NotEmpty
+    private String ip;
+    @NotEmpty
+    private String zookeepers;
+    @NotEmpty
+    private String instanceName;
+    @NotEmpty
+    private String username;
+    @NotEmpty
+    private String password;
+
+    private String table = "timely.metrics";
+    private String meta = "timely.meta";
+    private int metricAgeOffDays = 7;
+    private List<String> metricsReportIgnoredTags = new ArrayList<>();
+    private boolean allowAnonymousAccess = false;
+    private int sessionMaxAge = 86400;
+
+    @NestedConfigurationProperty
+    private Port ports = new Port();
+    @NestedConfigurationProperty
+    private Http http = new Http();
+    @NestedConfigurationProperty
+    private Write write = new Write();
+    @NestedConfigurationProperty
+    private Scanner scanner = new Scanner();
+    @NestedConfigurationProperty
+    private Cors cors = new Cors();
+    @NestedConfigurationProperty
+    private MetaCache metaCache = new MetaCache();
+    @NestedConfigurationProperty
+    private VisibilityCache visibilityCache = new VisibilityCache();
+    @NestedConfigurationProperty
+    private WebSocket webSocket = new WebSocket();
+    @NestedConfigurationProperty
+    private Ssl ssl = new Ssl();
+
+    public String getIp() {
+        return ip;
+    }
+
+    public void setIp(String ip) {
```

Comment:
we wire it up programmatically in the test cases. Perhaps we should create test config files or just not care.  The nested classes issue may also make it not any better than it is currently.